cloudmarker/test/test_filestore.py

```diff
@@ -0,0 +1,69 @@
+"""Tests for filestore plugins."""
+
+
+import os
+import pathlib
+import shutil
+import unittest
+
+from cloudmarker.stores import filestore
+
+
+class FileStoreTest(unittest.TestCase):
+    """Tests for filestore plugins."""
+
+    @classmethod
+    def write_to_file(cls, filename):
+        # Initialize the FileStore plugin with the filename
+        store = filestore.FileStore(filename)
+
+        # dummy data to write
+        mp = {
+            "record_type": "dummy",
+            "data": {
+                "test_name": "tilde_filename"
+            }
+        }
+
+        store.write(mp)
+        store.done()
+
+    def test_filepath_tilde(self):
+        self.write_to_file("~/caramel_l33t/cloudmarker")
+        # if the parsing fails, it should create a "~" directory in the current
+        # path
+        f = pathlib.Path("~")
+
+        # If exists and is a directory, the parsing failed.
+        self.assertFalse(f.exists() and f.is_dir())
+
+    def test_filepath_tilde2(self):
+        user = os.getenv("USER")
+        self.write_to_file("~{}/caramel_tpb/cloudmarker".format(user))
+        # if the parsing fails, it should create a "~" directory in the current
+        # path
+        f = pathlib.Path("~{}".format(user))
+
+        # If exists and is a directory, the parsing failed.
+        self.assertFalse(f.exists() and f.is_dir())
+
```

Comment:
I think it would be a more accurate to check for the existence of the file itself which would automatically check the whole path also. Something like `os.path.isfile("$HOME/caramel_tpb/cloudmarker/dummy.json"`. I know that the $HOME is only for Linux, might need something that would work for cross-platform paths. The condition here is minimalistic and beautiful in that it checks if the path was created in the wrong place, but it is not checking if the path expansion actually happened. Any views @prateeknischal , @susam ?

Reply:
@s-nirali I agree. Ideally, there should be additional tests to check the positive condition, i.e., that the file is written to the correct path. Such a test should be another separate test method. While such a test case would be nice, even if it is missing in this pull request (PR), we can still merge this PR with the current set of test methods after the review comments are addressed, and further tests may be added later by anyone.

By the way, a cross-platform way to get the home directory is `os.path.expanduser('~')`.

Reply:
I have added `pathlib.Path(os.path.join("/tmp/abc", "dummy.json"))` for testing the positive condition, but as discussed with @s-nirali, there is actually no need for this test file as we are now basically testing python code and not cloudmarker code.

To write a positive test for expanduser, I will have to use expanduser function, which does not make any sense. Would it be ok to remove all the tests and let just the have the test in #22 only.

Reply:
I think there should be at least one test for this. The fact that `FileStore` performs tilde expansion on the path is a Cloudmarker behavior. If someday in future, we inadvertently stop doing tilde expansion, then at least one test should break that should tell us that the expected behavior has gone missing.

Regarding invoking `os.path.expanduser()` to test something that does `os.path.expanduser()`, I agree that it seems self-referential indeed. However, at the same time it is also good to capture the expected behaviour as a test. At least, it would allow for things like changing the `FileStore` implementation from using `os.path.expanduser()` to something like `pathlib.Path(path).expanduser()` in future, say we don't want to support Python 3.4 anymore. The test would guarantee that nothing breaks during such a change.

This turns out to be one of those decisions that depends on one's judgement. I prefer having at least one test. I'll let you decide what you want to do with the tests. 🙂

By the way, the `pathlib` module is featureful enough to not require `os.path.join()`. For example, `pathlib.Path('/tmp/abc', 'dummy.json')` produces the same result as `pathlib.Path(os.path.join("/tmp/abc", "dummy.json"))`.